src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+  return { start, end };
+};
+
+// the range including the whole lines that the node is in
+const blockRange = (source: TextLines, node: commonmark.Node): Offset => {
+  const [[ startLine ], [ endLine ]] = node.sourcepos;
+  const start = posToOffset(source.lines, [ startLine, 1 ]);
+  const end = (endLine < source.lines.length ? posToOffset(source.lines, [ endLine + 1, 1 ]) : source.text.length) - 1;
+  return { start, end };
+};
+
+const extractText = (source: TextLines, node: commonmark.Node): string => {
+  const range = columnRange(source.lines, node);
+  return source.text.substring(range.start, range.end + 1);
+};
+
+const pos = (node: commonmark.Node): string => {
+  if (node.sourcepos !== undefined) {
+    const [[ startLine, startColumn ]] = node.sourcepos;
+    return ` (line: ${startLine} column: ${startColumn})`;
+  } else {
+    return '';
+  }
+};
+
+const parseHeadings = (changelog: string): Top => {
+  const data: Top = { preamble: [], subheadings: [] };
+  const reader = new commonmark.Parser();
+  const doc = reader.parse(changelog);
+  const stack: (Heading | Top)[] = [ data ];
+  let node: commonmark.Node | null = doc.firstChild;
+  while (node !== null) {
+    if (node.type === 'heading') {
+      // pop headings off the stack that are higher or equal level
+      while (true) {
+        const last = stack[stack.length - 1];
+        if (node.level <= (last?.header?.level ?? 0)) {
+          stack.pop();
+        } else {
+          break;
+        }
+      }
+      // push this as a new child heading
+      const heading: Heading = {
+        header: node,
+        preamble: [],
+        subheadings: []
+      };
+      stack[stack.length - 1].subheadings.push(heading);
+      stack.push(heading);
+    } else {
+      stack[stack.length - 1].preamble.push(node);
+    }
+    node = node.next;
+  }
+  return data;
+};
+
+const jiraRe = /\s+#(?<jira>[A-Z]{2,10}-\d+)\s*$/;
+const parseItem = (source: TextLines, node: commonmark.Node): E.Either<string[], Item> => {
+  if (node.type !== 'item') {
+    return E.left([ 'Expected a list item' + pos(node) ]);
+  }
+  const offset = blockRange(source, node);
+  const itemText = extractText(source, node);
+  const m = jiraRe.exec(itemText);
+  const jira = m?.groups?.jira;
+  return E.right({
+    ...offset,
+    jira
+  });
+};
+
+const parseList = (source: TextLines, list: commonmark.Node): E.Either<string[], Item[]> => {
+  const itemsE: E.Either<string[], Item>[] = [];
+  let item = list.firstChild;
+  while (item) {
+    itemsE.push(parseItem(source, item));
+    item = item.next;
+  }
+  return EitherUtils.combine(itemsE);
+};
+
+const sectionNames: SectionNames[] = [ 'Added', 'Improved', 'Changed', 'Deprecated', 'Removed', 'Fixed', 'Security' ];
+const sectionRes = sectionNames.map((sectionName) => new RegExp(`^### ${sectionName}$`));
+const parseFragment = (source: TextLines, sectionHeadings: Heading[]): E.Either<string[], ChangelogFragment> => {
```

Comment:
This function is very difficult to follow, could we maybe split it up into some smaller functions to better understand the flow? Even just a function or two to do all the validation and push to the errors would help